src/Field.tsx

```diff
@@ -190,4 +195,78 @@ class FieldInner<Values = {}, Props = {}> extends React.Component<
   }
 }
 
-export const Field = connect<FieldAttributes<any>, any>(FieldInner);
+export const Field = connect<FieldOuterProps<any>, any>(FieldInner);
+
+export type TypedAttributes<Values, ValueType> = Partial<
+  Pick<
+    FieldAttributes<Values, ValueType>,
+    Exclude<keyof FieldAttributes<Values, ValueType>, 'name'>
+  >
+>;
+export type WrapFieldFunction<
+  FormValues,
+  Parent,
+  Values,
+  Key extends keyof Values
+> = (
+  parent: Parent
+) => React.ComponentType<TypedAttributes<FormValues, Values[Key]>>;
+
+export interface FieldDefinition<FormValues, ValueType> {
+  _parent?: FieldDefinition<FormValues, any>;
+  _key?: string;
+  _field: React.ComponentType<TypedAttributes<FormValues, ValueType>>;
+}
+
+export type TypedFieldProxy<FormValues, Values = FormValues> = {
+  [fieldName in keyof Values]: Values[fieldName] extends object
+    ? TypedFieldProxy<FormValues, Values[fieldName]>
+    : FieldDefinition<FormValues, Values[fieldName]>
+} &
+  FieldDefinition<FormValues, any>;
+
+const wrapField = <FormValues, Values, Name extends keyof Values>(
+  key: Name,
+  parent?: FieldDefinition<FormValues, any>
+) => {
+  let path: string = key + ''; // stringify
+
+  while (parent && parent._key) {
+    path = `${parent._key!}[${path}]`;
+    parent = parent._parent;
+  }
+
+  return (props: TypedAttributes<FormValues, Values[Name]>) => {
+    return <Field name={path} {...props} />;
+  };
+};
+
+export const typedFieldProxy = <
+  FormValues,
+  Values = FormValues,
+  Parent extends FieldDefinition<FormValues, any> | undefined = undefined
+>(
+  parent?: Parent
+) => {
+  return new Proxy({} as TypedFieldProxy<FormValues, Values>, {
```

Comment:
Can we dogfood this in userland without making changes to Formik? eg. make `import { Field } from 'typed-field'`?

Reply:
I think the only way for that to make sense would be to separate the entire type definition from this project into `typed-formik`. I would not want to maintain a separate strongly typed definition of the fields that will break the instant you push 2.0. What I've done is strengthen the types of `<Field />` itself while leaving the default export `Field` to function exactly the way it does currently, seamlessly integrating it into the Formik API in a way for any user to opt-in to the types already defined by Formik, providing the functionality that is suggested by https://jaredpalmer.com/formik/docs/guides/typescript .

Reply:
Cancel the old response. I think I'd be okay with importing a Formik wrapper that provides `Fields` separately, if you are concerned about maintaining that. However, in order for that to happen, we'd still need to commit to a set of real typings for fields. If we cannot strongly type the fields themselves and commit to it, then this will be better as a fork.